### DISCUSSION TITLE
missingSuspenseWithCSRBailout in Next.js v14.1 with App Router

### DISCUSSION THREAD

**Discussion by ajaymahadeven** [vercel/next.js#61298] [Help]
### Summary

During the build process of a Next.js application, I encountered warnings indicating that entire pages are being defaulted to client-side rendering due to the use of `useSearchParams()`. The warnings persist despite not using `useSearchParams()` in the mentioned pages. I've attempted to mitigate this issue by enabling `missingSuspenseWithCSRBailout: false` in my `next.config.js`, but the problem remains unresolved. I'm seeking advice on how to address these warnings and ensure proper rendering.

Currently, I'm avoiding this problem by using :
```
experimental: {
  missingSuspenseWithCSRBailout: false,
}
```
as mentioned in the official documentation. Additionally, I'm using Clerk.js for authentication and user management.

### Additional information

```js
Here's an example of the error message I'm receiving for multiple pages, including "/404", "/dashboard", "/", "/forms", "/notes", and "/settings":

 ⚠ Entire page "/404" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

 ⚠ Entire page "/dashboard" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

 ⚠ Entire page "/" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

 ⚠ Entire page "/forms" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

 ⚠ Entire page "/notes" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

 ⚠ Entire page "/settings" deopted into client-side rendering due to "useSearchParams()". Read more: https://nextjs.org/docs/messages/deopted-into-client-rendering

Example of forms page code (as of writing):


"use client";
import React, { Suspense } from "react";

import NavBar from "@/components/nav-bar";
import Banner from "@/components/ui/banner";

export default function Page() {
  return (
    <Suspense>
      <main className="flex flex-col  h-dvh">
        <NavBar />
        <>
          <Banner bannerTitle={"Forms"} />
        </>
      </main>
    </Suspense>
  );
}
```

**Comment by icyJoseph** [comment]
Hi,

Are you using `useNextRouter`, from Clerk? That one uses `useSearchParams`:

- https://github.com/clerk/javascript/blob/da4090fd7e0a5926589f71b17ec4978f33a455fa/packages/elements/src/react/router/next.ts#L2

Although the flag you link to is experimental, I am surprised that it doesn't do the trick.

```js
// next.config.js
module.exports = {
   experimental: {
      missingSuspenseWithCSRBailout: false
   }
}
```

This, should've been all that's needed, according to:

- https://github.com/vercel/next.js/blob/93fb762fc66a0fe2d53790fda0d89825fa50bd06/packages/next/src/server/app-render/app-render.tsx#L1017

**Reply by ajaymahadeven** [reply]
Hey @icyJoseph, thanks for getting back to me so quickly. I wanted to share that I'm not utilizing `useNextRouter` from Clerk. Instead, I've implemented the following Clerk.js functionalities in my project:

**Middleware Configuration:**

I've set up the middleware using `authMiddleware` from `@clerk/nextjs` to protect all routes, including API and TRPC routes. The configuration allows specifying public routes. For more details on customizing the middleware, you can refer to the Clerk documentation on auth middleware.

```javascript
import { authMiddleware } from "@clerk/nextjs";

export default authMiddleware({
  publicRoutes: ["/", "/terms", "/privacy"],
});

export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

**User Context Provider:**

For managing user state, I've created a `UserProvider` component using the `useUser` hook from `@clerk/clerk-react`. This provider component makes user information, such as the first name and last login time, accessible throughout the application.

```javascript
"use client";
import React, { createContext, useContext } from "react";
import { useUser } from "@clerk/clerk-react";

interface UserContextProps {
  user: object | null | undefined;
  firstName: string;
  lastLogin: string;
  isLoaded: boolean;
}

const UserContext = createContext<UserContextProps | null>(null);

export function UserProvider({ children }: { children: React.ReactNode }) {
  const { user } = useUser();

  const firstName: string = user?.firstName?.split(" ")[0] || "Unknown User";
  const lastLogin = useUser().user?.lastSignInAt?.toLocaleString() || "Unknown Date";
  const isLoaded = useUser().isLoaded;

  return (
    <UserContext.Provider value={{ user, firstName, lastLogin, isLoaded }}>
      {children}
    </UserContext.Provider>
  );
}

export function useUserContext() {
  return useContext(UserContext);
}
```

In other parts of the application, I simply reference the user's logged-in status, their first name, and the last login time directly from this provider.

I hope this clarifies how I've integrated Clerk.js into my project. Let me know if you have any further questions or insights!

**Reply by icyJoseph** [reply]
Does your layout or any other part use `useSearchParams` at all?

**Reply by ajaymahadeven** [reply]
Hey @icyJoseph, I've shared the core code of my current layout for your review. As you'll notice, I'm not utilising `useSearchParams` in any part of this setup, which is why I reported this and it's quite bizarre as well. 

```javascript
import type { Metadata } from "next";
import { Inter } from "next/font/google";

import { Providers } from "@/providers/providers";

import "./globals.css";

// Setting up the Inter font
const inter = Inter({ subsets: ["latin"] });

// RootLayout component definition
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en" suppressHydrationWarning>
      <body className={inter.className}>
        <Providers>{children}</Providers>
      </body>
    </html>
  );
}
```

This snippet represents the minimal foundational code of my layout as it stands.

**Reply by alejandro0619** [reply]
experimenting the same problem, any news on that?

**Reply by vignesh-gupta** [reply]
Hey @icyJoseph, question would be why the flag in config is not recognizable.

Here's the error:
```
 ⚠ Invalid next.config.mjs options detected: 
 ⚠     Unrecognized key(s) in object: 'missingSuspenseWithCSRBailout' at "experimental"
 ⚠ See more info here: https://nextjs.org/docs/messages/invalid-next-config
```

**Reply by VanTyse** [reply]
I noticed that you have a Providers component. You may want to check all your providers, maybe comment them out one by one until you discover which of them is causing the issue.

**Reply by yackermann** [reply]
Same here
